pkg/schedule/checker/affinity_checker.go

```diff
@@ -0,0 +1,434 @@
+	"github.com/tikv/pd/pkg/schedule/affinity"
+	"github.com/tikv/pd/pkg/schedule/config"
+	sche "github.com/tikv/pd/pkg/schedule/core"
+	"github.com/tikv/pd/pkg/schedule/filter"
+	"github.com/tikv/pd/pkg/schedule/operator"
+	"github.com/tikv/pd/pkg/schedule/placement"
+	"github.com/tikv/pd/pkg/schedule/types"
+	"github.com/tikv/pd/pkg/utils/logutil"
+)
+
+// AffinityChecker groups regions with affinity labels together by affinity group.
+// It ensures regions adhere to affinity group constraints by creating operators.
+type AffinityChecker struct {
+	PauseController
+	cluster         sche.CheckerCluster
+	affinityManager *affinity.Manager
+	conf            config.CheckerConfigProvider
+}
+
+// NewAffinityChecker create an affinity checker.
+func NewAffinityChecker(cluster sche.CheckerCluster, conf config.CheckerConfigProvider) *AffinityChecker {
+	return &AffinityChecker{
+		cluster:         cluster,
+		affinityManager: cluster.GetAffinityManager(),
+		conf:            conf,
+	}
+}
+
+// GetType return AffinityChecker's type.
+func (*AffinityChecker) GetType() types.CheckerSchedulerType {
+	return types.AffinityChecker
+}
+
+// Name returns AffinityChecker's name.
+func (*AffinityChecker) Name() string {
+	return types.AffinityChecker.String()
+}
+
+// Check verifies a region's replicas according to affinity group constraints, creating an Operator if needed.
+func (c *AffinityChecker) Check(region *core.RegionInfo) []*operator.Operator {
+	if c.affinityManager == nil {
+		return nil
+	}
+	affinityCheckerCounter.Inc()
+
+	if c.IsPaused() {
+		affinityCheckerPausedCounter.Inc()
+		return nil
+	}
+
+	// Check region state
+	if region.GetLeader() == nil {
+		affinityCheckerRegionNoLeaderCounter.Inc()
+		return nil
+	}
+	if !filter.IsRegionHealthy(region) {
+		affinityCheckerUnhealthyRegionCounter.Inc()
+		return nil
+	}
+	if !filter.IsRegionReplicated(c.cluster, region) {
+		affinityCheckerAbnormalReplicaCounter.Inc()
+		return nil
+	}
+
+	// Get the affinity group for this region
+	group, isAffinity := c.affinityManager.GetRegionAffinityGroupState(region)
+	if group == nil {
+		// Region doesn't belong to any affinity group
+		return nil
+	}
+
+	// In some cases, the Group must be updated via ObserveAvailableRegion and then fetched again.
+	needRefetch := false
+	if group.AffinitySchedulingAllowed {
+		// If the Group is affinity scheduling allowed and the Region is not in the affinity state,
+		// we expect to schedule the Region to match the Group’s peers.
+		// Before doing so, check whether the Group’s peers are replicated.
+		// If not, the Group information should be expired (e.g. due to Placement Rules changes),
+		// so expire the group first, then provide the available Region information and fetch the Group state again.
+		if !isAffinity {
+			targetRegion := cloneRegionWithReplacePeerStores(region, group.LeaderStoreID, group.VoterStoreIDs...)
+			if targetRegion == nil || !filter.IsRegionReplicated(c.cluster, targetRegion) {
```

Comment:
Only check replicated?

Reply:
The health check for the store is performed in the `checkStoresAvailability` function.